In [5]:
import pandas as pd
import numpy as np

from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import classification_report

### 1. Create a new file named `model_evaluation.py` or `model_evaluation.ipynb` for these exercises.
### 2. Given the following confusion matrix, evaluate (by hand) the model's performance
**Assuming we are predicing dog, "It will be dog" is the positive prediction.**

In [3]:
confusion_matrix = pd. DataFrame([[34, 7], [13, 46]], 
                                 index=['predicted cat', 'predicted dog'], 
                                 columns=['actual cat', 'actual dog'])
confusion_matrix.T

predicted cat  predicted dog
actual cat             34             13
actual dog              7             46

### 2-a. In the context of this problem, what is a false positive?
The false positive is that the model is predicting a dog but it is actually a cat. 

### 1-b. In the context of this problem, what is a false negative?
The false negative is that the model is predicting a cat but it is actually a dog. 

### 1-c. How would you describe this model? 

In [4]:
print(f"The accuracy of the model is: {(34+46)/100}")
print(f"The precision of the model is: {46/(46+13)}")
print(f"The recall of the model is: {46/(46+7)}")
print(f"The specificity of the model is: {34/(34+13)}")

The accuracy of the model is: 0.8
The precision of the model is: 0.7796610169491526
The recall of the model is: 0.8679245283018868
The specificity of the model is: 0.723404255319149


### 2. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant. Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here. Use the predictions dataset and pandas to help answer the following questions.

In [6]:
df_duck = pd.read_csv('c3.csv')
df_duck.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [58]:
# df_duck.actual = df_duck.actual.map({'No Defect': 1, 'Defect': 0})
# df_duck.model1 = df_duck.model1.map({'No Defect': 1, 'Defect': 0})
# df_duck.model2 = df_duck.model2.map({'No Defect': 1, 'Defect': 0})
# df_duck.model3 = df_duck.model3.map({'No Defect': 1, 'Defect': 0})

In [7]:
df_duck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


In [8]:
df_duck.isnull().any()

actual    False
model1    False
model2    False
model3    False
dtype: bool

In [9]:
df_duck.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

**Takeways**:
1. The most common class in actual is 'No Defect'.
2. 'No Defect' is the postive prediction. 
3. 'Defect' is the negative prediction.
4. FP: predicting 'No Defect' but actually have 'Defect'. 
5. FN: predicting 'Defect' but actually 'No Defect'. 

### 2-a. An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

* They want the TN: predicted Defect(N) and actually Defect(N).
* They don't want the FP: predicted No Defect(P) but actually Defect (N). 
* `specificity` would be appropriate here: TN / TN + FP.
* The model with the highest `specificity` will be the best fit. 

In [82]:
cross_tab_model_1 = pd.crosstab(df_duck.actual, df_duck.model1)
cross_tab_model_1

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [75]:
cross_tab_model_2 = pd.crosstab(df_duck.actual, df_duck.model2)
cross_tab_model_2

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [76]:
cross_tab_model_3 = pd.crosstab(df_duck.actual, df_duck.model3)
cross_tab_model_3

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

In [77]:
specificity_model_1 = 8/16
specificity_model_2 = 9/16
specificity_model_3 = 13/16

specificity_model_1, specificity_model_2, specificity_model_3

(0.5, 0.5625, 0.8125)

In [81]:
df_duck.model1.value_counts()

No Defect    190
Defect        10
Name: model1, dtype: int64

In [91]:
# Recall of the negative class is "specificity"

print(classification_report(df_duck.actual, df_duck.model1))
print(classification_report(df_duck.actual, df_duck.model2))
print(classification_report(df_duck.actual, df_duck.model3))

              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200

              precision    recall  f1-score   support

      Defect       0.10      0.56      0.17        16
   No Defect       0.94      0.56      0.70       184

    accuracy                           0.56       200
   macro avg       0.52      0.56      0.44       200
weighted avg       0.87      0.56      0.66       200

              precision    recall  f1-score   support

      Defect       0.13      0.81      0.23        16
   No Defect       0.97      0.53      0.69       184

    accuracy                           0.56       200
   macro avg       0.55      0.67      0.46       200
weighted avg       0.90      0.56      0.65       200



**Takeaways**:
Model 3 has the highest specificity so it is the best fit. 

### 2-b. Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

* They need to predict TN (predicted N and acurally N)
* They don't want FN: predicted a Defect(N) but acurally No Defect(P). 
* Hence, `precision(N)` should be used as the evaluation metric here. 
* The model with the highest `precision(N)` will be the best fit. 
* Precision(N) = TN / (TN + FN)

In [11]:
# Precision(N) is written as 'P_N'

P_N_model_1 = 8/(8+2)
P_N_model_2 = 9/(9+81)
P_N_model_3 = 13/(13+86)

P_N_model_1, P_N_model_2, P_N_model_3

(0.8, 0.1, 0.13131313131313133)

In [92]:
# Recall of the negative class is "specificity"

print(classification_report(df_duck.actual, df_duck.model1))
print(classification_report(df_duck.actual, df_duck.model2))
print(classification_report(df_duck.actual, df_duck.model3))

              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200

              precision    recall  f1-score   support

      Defect       0.10      0.56      0.17        16
   No Defect       0.94      0.56      0.70       184

    accuracy                           0.56       200
   macro avg       0.52      0.56      0.44       200
weighted avg       0.87      0.56      0.66       200

              precision    recall  f1-score   support

      Defect       0.13      0.81      0.23        16
   No Defect       0.97      0.53      0.69       184

    accuracy                           0.56       200
   macro avg       0.55      0.67      0.46       200
weighted avg       0.90      0.56      0.65       200



**Takeways**:
Model 3 has the highest precision so it is the best fit for this use case

### 3. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. 

1. First an automated algorithm tags pictures as either a cat or a dog (Phase I). 
2. Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here. Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions.

In [12]:
df = pd.read_csv('gives_you_paws.csv')
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  5000 non-null   object
 1   model1  5000 non-null   object
 2   model2  5000 non-null   object
 3   model3  5000 non-null   object
 4   model4  5000 non-null   object
dtypes: object(5)
memory usage: 195.4+ KB


In [14]:
df.isnull().any()

actual    False
model1    False
model2    False
model3    False
model4    False
dtype: bool

In [17]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [18]:
# Creat the baseline model according to the actual observations

df['baseline'] = df.actual.value_counts().index[0]
df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [20]:
(df.model1 == df.actual).mean()

0.8074

### 3-a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

* Accuracy = TP + TN / All
* accuracy = (df.prediction == df.actual).mean()

In [31]:
# Get all model columns

models = df.loc[:, 'model1':'baseline'].columns.tolist()
models

['model1', 'model2', 'model3', 'model4', 'baseline']

In [38]:
output = []

for model in models:
    output.append({
        "model": model,
        "accuracy": (df[model] == df.actual).mean()
    })

metrics = pd.DataFrame(output)
metrics = metrics.sort_values(by='accuracy', ascending=False, ignore_index=True)
metrics

model  accuracy
0    model1    0.8074
1    model4    0.7426
2  baseline    0.6508
3    model2    0.6304
4    model3    0.5096

In [7]:
cross_table_model_1 = pd.crosstab(df.actual, df.model1)
cross_table_model_1

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

In [8]:
cross_table_model_2 = pd.crosstab(df.actual, df.model2)
cross_table_model_2

model2   cat   dog
actual            
cat     1555   191
dog     1657  1597

In [9]:
cross_table_model_3 = pd.crosstab(df.actual, df.model3)
cross_table_model_3

model3   cat   dog
actual            
cat      893   853
dog     1599  1655

In [10]:
cross_table_model_4 = pd.crosstab(df.actual, df.model4)
cross_table_model_4

model4  cat   dog
actual           
cat     603  1143
dog     144  3110

In [37]:
baseline_model = 3254/5000
accuracy_model_1 = (1423+2614)/5000
accuracy_model_2 = (1555+1597)/5000
accuracy_model_3 = (893+1655)/5000
accuracy_model_4 = (603+3110)/5000

baseline_model, accuracy_model_1, accuracy_model_2, accuracy_model_3, accuracy_model_4

(0.6508, 0.8074, 0.6304, 0.5096, 0.7426)

In [45]:
accuracy_model_1 = accuracy_score(df.actual, df.model1)
accuracy_model_2 = accuracy_score(df.actual, df.model2)
accuracy_model_3 = accuracy_score(df.actual, df.model3)
accuracy_model_4 = accuracy_score(df.actual, df.model4)

accuracy_model_1, accuracy_model_2, accuracy_model_3, accuracy_model_4

(0.8074, 0.6304, 0.5096, 0.7426)

**Takeways**:
1. Model 1 and Model 4 are better than baseline in term of accuracy
2. Model 2 and Model 3 are worse than baseline in term of accuracy

**Takeaways**:
1. The most common class is dog
2. Predicing dog will be the postivie direction
3. TP = 3254, TN = 0, FP = 1746, FN = 0

### 3-b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

* Since Phase I is to tags pictures as either a cat or a dog, so accuracy i used as the evluation metrics. 
1. The team only wants dog pictures
2. False Negtaive is more tolerated than False Positive
3. Precision is the evalution metric for both Phase I and II. 

In [ ]:
precision_model_1 = 2614/(2614+323)
precision_model_2 = 1597/(1597+191)
precision_model_3 = 1655/(1655+853)
precision_model_4 = 3110/(3110+1143)

precision_model_1**2, precision_model_2**2, precision_model_3**2, precision_model_

**Takeaways**

I will recommend model 2. 

### 3-c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

1. The team only wants cat pictures.
2. Specificty is the evalution metric for both Phase I and II

In [95]:
specificity_model_1 = 1423/(1423+323)
specificity_model_2 = 1555/(1555+191)
specificity_model_3 = 893/(893+853)
specificity_model_4 = 603/(603+1143)

specificity_model_1, specificity_model_2, specificity_model_3, specificity_model_4

(0.8150057273768614,
 0.8906071019473081,
 0.5114547537227949,
 0.34536082474226804)

In [97]:
print(classification_report(df.actual, df.model1))
print(classification_report(df.actual, df.model2))
print(classification_report(df.actual, df.model3))
print(classification_report(df.actual, df.model4))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000

              preci

**Takeaways**

I will recommend model 2. 

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

Please find the answers within Q2 and Q3.